custom/rch/views.py

```diff
@@ -0,0 +1,335 @@
+
+    def _set_context_for_matched_records(self, context, in_data=None):
+        beneficiary_type = in_data.get('beneficiary_type')
+        if beneficiary_type == 'child':
+            self.beneficiaries = RCHChildRecord.objects
+        else:
+            self.beneficiaries = RCHMotherRecord.objects
+
+        self.beneficiaries = self.beneficiaries.exclude(cas_case_id__isnull=True)
+        size = in_data.get('limit')
+        offset = (in_data.get('page') - 1) * in_data.get('limit')
+        if self._set_rch_beneficiaries(True, only_matched=True, in_data=in_data):
+            context['beneficiaries_total'] = self.beneficiaries.count()
+            context['beneficiaries'] = self.beneficiaries[offset:offset + size]
+
+    def _set_beneficiaries(self, context, in_data):
+        context['beneficiaries'] = []
+        present_in = in_data.get('present_in')
+        if present_in == 'cas':
+            self._set_context_for_cas_records(context, in_data)
+        elif present_in == 'rch':
+            self._set_context_for_rch_records(context, in_data)
+        elif present_in == 'both':
+            self._set_context_for_matched_records(context, in_data)
+
+    def get_context_data(self, **kwargs):
+        context = super(BeneficariesList, self).get_context_data(**kwargs)
+        context['filter_form'] = BeneficiariesFilterForm(data=self.request.GET, domain=self.request.domain)
+        context['pagination_limit_cookie_name'] = ('hq.pagination.limit.rch_cas_beneficiaries.%s'
+                                                   % self.request.domain)
+        return context
+
+    def get_template_names(self):
+        if self.request.GET.get('present_in') == 'cas':
+            return "cas/beneficiaries_list.html"
+        return self.template_name
+
+    @staticmethod
+    def _mask_aadhar_number(beneficiary_json):
+        for aadhar_num_field in AADHAR_NUM_FIELDS:
+            if beneficiary_json.get(aadhar_num_field):
+                beneficiary_json[aadhar_num_field] = mask_aadhar_number(beneficiary_json[aadhar_num_field])
+
+    def _format_beneficiary(self, beneficiary):
+        beneficiary_json = json.loads(serializers.serialize('json', [beneficiary]))[0].get('fields')
+        beneficiary_json['editUrl'] = reverse(
+            BeneficiaryView.urlname,
+            args=[ICDS_CAS_DOMAIN,
+                  beneficiary.beneficiary_type,
+                  beneficiary.id])
+        self._mask_aadhar_number(beneficiary_json)
+        return beneficiary_json
+
+    @allow_remote_invocation
+    def get_pagination_data(self, in_data):
+        context = {}
+        if in_data.get('present_in'):
+            self._set_beneficiaries(context, in_data)
+
+        if in_data.get('present_in') == 'cas':
+            itemList = context.get('beneficiaries')
+            for item in itemList:
+                item['editUrl'] = reverse(CaseDetailsView.urlname, args=(ICDS_CAS_DOMAIN, item.get('_id')))
```

Comment:
can this not be done in JS? Doesn't seem great to do essentially the same reverse for every item.
```
# in python
context['case_details_view_url'] = reverse(CaseDetailsView.urlname, args=[request.domain, '---'])

# in template
{% initial_page_data 'case_details_view_url' case_details_view_url %}

# in JS
var url = hqImport('hqwebapp/js/initial_page_data').reverse('case_details_view_url', item._id)
```

Reply:
Same in `_format_beneficiary`